### ISSUE TITLE
Meaningless operations can lead to inconsistent performance

### ISSUE THREAD

**Issue by MTandHJ** [pyg-team/pytorch_geometric#5838]
### 🐛 Describe the bug

```
import torch
from torch_geometric.nn import LGConv

class LightGCN(torch.nn.Module):

    def __init__(
        self, tokenizer: Tokenizer, 
        graph: Data,
        num_layers: int = 3
    ) -> None:
        super().__init__()

        self.tokenizer = tokenizer
        self.conv = LGConv()
        self.num_layers = num_layers
        self.graph = graph

        self.User, self.Item = self.tokenizer[USER, ID], self.tokenizer[ITEM, ID]

        self.initialize()

    @property
    def graph(self):
        return self.__graph

    @graph.setter
    def graph(self, graph: Data):
        self.__graph = graph
        T.ToSparseTensor()(self.__graph)

    def to(
        self, device: Optional[Union[int, torch.device]] = None, 
        dtype: Optional[Union[torch.dtype, str]] = None, 
        non_blocking: bool = False
    ):
        if device:
            self.graph.to(device)
        return super().to(device, dtype, non_blocking)

    def forward(
        self, users: Optional[Dict[str, torch.Tensor]] = None, 
        items: Optional[Dict[str, torch.Tensor]] = None
    ):
        userEmbs = self.User.embeddings.weight
        itemEmbs = self.Item.embeddings.weight
        features = torch.cat((userEmbs, itemEmbs), dim=0).flatten(1) # N x D
        avgFeats = features / (self.num_layers + 1)
        for _ in range(self.num_layers):
            features = self.conv(features, self.graph.adj_t)
            avgFeats += features / (self.num_layers + 1)
        userFeats, itemFeats = torch.split(avgFeats, (self.User.count, self.Item.count))

        if self.training: # Batch
            users, items = users[self.User.name], items[self.Item.name]
            userFeats = userFeats[users] # B x 1 x D
            itemFeats = itemFeats[items] # B x n x D
            userEmbs = self.User.look_up(users) # B x 1 x D
            itemEmbs = self.Item.look_up(items) # B x n x D
            return torch.mul(userFeats, itemFeats).sum(-1), userEmbs, itemEmbs
        else:
            return userFeats, itemFeats
```

I tried to reproduce LightGCN on Gowalla dataset using codes above. However, it did not work well all the time.
Sometimes the loss starts from `0.68` and struggles to decrease. Unexpectedly, without changing any code but adding a line of `print("???")` at the beginning will make it work again; that is, the loss decreases from `0.44` and finally achieves NDCG@20: 0.1540. Moreover, `seed` is set by

```
def set_seed(seed: int) -> int:
    """Set seed.

    Parameters:
    seed: int
        - `-1`: Set a seed from 0 to 2048 randomly.
        - `int`: Set a seed of `int`.

    Returns:
    seed: int
    """
    if seed == -1:
        seed = random.randint(0, 2048)
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    return seed
```
 

I also observe the same loss in the first few mini-batches but without the `nan` values during training. Is there anything wrong with my implementation?

Thanks !

### Environment

* PyG version:  2.1.0
* PyTorch version: 1.12.1
* OS: Ubuntu 16.04.5
* Python version: 3.9
* CUDA/cuDNN version: 10.2/10.2
* How you installed PyTorch and PyG (`conda`, `pip`, source): conda

**Comment by MTandHJ** [comment]
I also observe different scores in the first mini-batches:

without `print("???")`:

    tensor([ 1.0588e-09, -7.0399e-09], device='cuda:0', grad_fn=<SelectBackward0>)

with `print("???")`:

    tensor([ 2.5725e-10, -1.0834e-08], device='cuda:0', grad_fn=<SelectBackward0>)

**Comment by MTandHJ** [comment]
I have found the culprit causing this problem.  I construct a graph from heteograph like that:

```
   @timemeter("DataSet/to_graph")
    def to_heterograph(self, *edge_types: Tuple[SparseField, Optional[str], SparseField]) -> HeteroData:
        """Convert datapipe to a heterograph.

        Parameters:
        ---

        *edge_types: (src, edge, dst)
            - src: SparseField
                Source node.
            - edge: Optional[str]
                The name of the edge. `src.name2dst.name` will be specified if `edge` is `None`.
            - dst: SparseField
                Destination node.

        Notes:
        ---

        Warning will be raised if current mode is not 'train' !

        Examples:
        ---

        >>> from freerec.data.datasets import MovieLens1M
        >>> from freerec.data.tags import USER, ITEM, ID
        >>> basepipe = MovieLens1M("../data/MovieLens1M")
        >>> fields = basepipe.fields
        >>> graph = basepipe.to_heterograph(
        ...    (fields[USER, ID], None, fields[ITEM, ID]), 
        ...    (fields[ITEM, ID], None, fields[USER, ID])
        ... )
        >>> graph
        HeteroData(
            UserID={ x=[6040, 0] },
            ItemID={ x=[3706, 0] },
            (UserID, UserID2ItemID, ItemID)={ edge_index=[2, 994169] },
            (ItemID, ItemID2UserID, UserID)={ edge_index=[2, 994169] }
        )
        """
        if self.mode != 'train':
            warnLogger(f"Convert the datapipe for {self.mode} to graph. Ensure this is intentional ...")

        srcs, _, dsts = zip(*edge_types)
        edges = list(map(lambda triplet: triplet[1] if triplet[1] else f"{triplet[0].name}2{triplet[2].name}", edge_types))
        nodes = srcs + dsts
        data = {node.name: [] for node in nodes}
        for chunk in self:
            for node in data:
                data[node].append(np.ravel(chunk[node]))
        for key in data:
            data[key] = torch.tensor(np.concatenate(data[key], axis=0), dtype=torch.long)

        graph = HeteroData()
        for node in nodes:
            graph[node.name].x = torch.empty((node.count, 0), dtype=torch.long)
        for src, edge, dst in zip(srcs, edges, dsts):
            u, v = data[src.name], data[dst.name]
            graph[src.name, edge, dst.name].edge_index = torch.stack((u, v), dim=0) # 2 x N
        print(srcs[0].name, edges, dsts[0].name)
        return graph.coalesce()
```

and transform it to graph:

```
graph = self.to_heterograph((src, None, dst)).to_homogeneous()
graph.edge_index = to_undirected(graph.edge_index)
```


Due to the randomness of `set()`,  the heterograph could be in the form of

```
HeteroData(
  UserID={ x=[29858, 0] },
  ItemID={ x=[40981, 0] },
  (UserID, UserID2ItemID, ItemID)={ edge_index=[2, 810128] }
)
```

```
HeteroData(
  ItemID={ x=[40981, 0] },
  UserID={ x=[29858, 0] },
  (UserID, UserID2ItemID, ItemID)={ edge_index=[2, 810128] }
)
```

However, the `to_homogeneous` reorders the indices according to nodes but not edges.

**Comment by rusty1s** [comment]
Thanks for this issue. Can you clarify what you mean by "`to_homogeneous()` reorders the indices according to nodes but not edges"? Are you referring to a bug in the underlying implementation?

**Comment by MTandHJ** [comment]
> Thanks for this issue. Can you clarify what you mean by "`to_homogeneous()` reorders the indices according to nodes but not edges"? Are you referring to a bug in the underlying implementation?

It is not a bug but an unexpected case. Calling `to_homogeneous()` for

```
HeteroData(
  ItemID={ x=[40981, 0] },
  UserID={ x=[29858, 0] },
  (UserID, UserID2ItemID, ItemID)={ edge_index=[2, 810128] }
)
```

will get nodes `ItemID + UserID` instead of `UserID + ItemID`.  I think the latter case is more reasonable since `UserID` is the source node.

**Comment by rusty1s** [comment]
Currently, we give node types depending on the order of node types in `HeteroData`. I think we cannot really rely on edge type information here since most of the time, edge types are defined for both directions.

**Comment by MTandHJ** [comment]
I see. Thanks for your quick reply.
